File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque

def solution(n, computers):
    visited = [False] * n
    
    answer = 0
    
    for i in range(n):
        if not visited[i]:
            answer += 1
            queue = deque([i])
            visited[i] = True
            
            while queue :
                node = queue.popleft()
                for j in range(n):
                    if not visited[j] and node != j and computers[node][j] == 1:
                        visited[j] = True
                        queue.append(j)
    return answer    
            
                    
        
                

File: 프로그래머스/3/43162. 네트워크/네트워크.js
function solution(n, computers) {
  const adjList = {};

  computers.forEach((connection, currentComputer) => {
    connection.forEach((element, index) => {
      if (element === 0) {
        return;
      }
      if (!adjList[currentComputer]) {
        adjList[currentComputer] = [];
      }
      if (index !== currentComputer) {
        adjList[currentComputer].push(index);
      }
    });
  });

  const visited = new Set();

  const dfs = (node, visited) => {
    if (visited.has(node)) {
      return;
    }

    visited.add(node);
    (adjList[node] ?? []).forEach((neighbor) => dfs(neighbor, visited));
  };

  let answer = 0;
  for (let i = 0; i < n; i++) {
    if (!visited.has(i)) {
      dfs(i, visited);
      answer += 1;
    }
  }
  return answer;
}